In [1]:
#import pandas library
import pandas as pd

In [2]:
#import the database by using pandas function "read_csv"
df_train = pd.read_csv('train.csv')
df_train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

When reviewing the database, it was decided to divide it into 4 dataframes to be able to check each variable in more detail.

In [3]:
# Use of iloc function to divide the dataframe
df1 = df_train.iloc[:, 0:20]
df1.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  
0       Norm     1Fam     2Story            7            5       2003  
1       Norm     1Fam     1Story            6            8       1976  
2       Norm     1Fam     2Story            7            5       2001  
3       Norm     1Fam     2Story            7            5       1915  
4       Norm     1Fam     2Story            8            5       2000

In [4]:
df2 = df_train.iloc[:, 21:41]
df2.head()

RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual  \
0     Gable  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd   
1     Gable  CompShg     MetalSd     MetalSd       None         0.0        TA   
2     Gable  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd   
3     Gable  CompShg     Wd Sdng     Wd Shng       None         0.0        TA   
4     Gable  CompShg     VinylSd     VinylSd    BrkFace       350.0        Gd   

  ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  \
0        TA      PConc       Gd       TA           No          GLQ   
1        TA     CBlock       Gd       TA           Gd          ALQ   
2        TA      PConc       Gd       TA           Mn          GLQ   
3        TA     BrkTil       TA       Gd           No          ALQ   
4        TA      PConc       Gd       TA           Av          GLQ   

   BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating  \
0         706          Unf           0        150          856    GasA   
1         978          Unf           0        284         1262    GasA   
2         486          Unf           0        434          920    GasA   
3         216          Unf           0        540          756    GasA   
4         655          Unf           0        490         1145    GasA   

  HeatingQC  
0        Ex  
1        Ex  
2        Ex  
3        Gd  
4        Ex

In [5]:
df3 = df_train.iloc[:,42:62]
df3.head()

Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0      SBrkr       856       854             0       1710             1   
1      SBrkr      1262         0             0       1262             0   
2      SBrkr       920       866             0       1786             1   
3      SBrkr       961       756             0       1717             1   
4      SBrkr      1145      1053             0       2198             1   

   BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  \
0             0         2         1             3             1          Gd   
1             1         2         0             3             1          TA   
2             0         2         1             3             1          Gd   
3             0         1         0             3             1          Gd   
4             0         2         1             4             1          Gd   

   TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt  \
0             8        Typ           0         NaN     Attchd       2003.0   
1             6        Typ           1          TA     Attchd       1976.0   
2             6        Typ           1          TA     Attchd       2001.0   
3             7        Typ           1          Gd     Detchd       1998.0   
4             9        Typ           1          TA     Attchd       2000.0   

  GarageFinish  GarageCars  
0          RFn           2  
1          RFn           2  
2          RFn           2  
3          Unf           3  
4          RFn           3

In [6]:
df4 = df_train.iloc[:,63:]
df4.head()

GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  \
0         TA         TA          Y           0           61              0   
1         TA         TA          Y         298            0              0   
2         TA         TA          Y           0           42              0   
3         TA         TA          Y           0           35            272   
4         TA         TA          Y         192           84              0   

   3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  \
0          0            0         0    NaN   NaN         NaN        0       2   
1          0            0         0    NaN   NaN         NaN        0       5   
2          0            0         0    NaN   NaN         NaN        0       9   
3          0            0         0    NaN   NaN         NaN        0       2   
4          0            0         0    NaN   NaN         NaN        0      12   

   YrSold SaleType SaleCondition  SalePrice  
0    2008       WD        Normal     208500  
1    2007       WD        Normal     181500  
2    2008       WD        Normal     223500  
3    2006       WD       Abnorml     140000  
4    2008       WD        Normal     250000

After dividing the database we had to analyze what type of data we have to work with, if they are interger, floating point, boolean or string. We also need to know if there are missing values ​​among our variables

In [7]:
#We make use of the info function to have a  little bit of knowledge about our dataframes
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            1460 non-null   int64  
 1   MSSubClass    1460 non-null   int64  
 2   MSZoning      1460 non-null   object 
 3   LotFrontage   1201 non-null   float64
 4   LotArea       1460 non-null   int64  
 5   Street        1460 non-null   object 
 6   Alley         91 non-null     object 
 7   LotShape      1460 non-null   object 
 8   LandContour   1460 non-null   object 
 9   Utilities     1460 non-null   object 
 10  LotConfig     1460 non-null   object 
 11  LandSlope     1460 non-null   object 
 12  Neighborhood  1460 non-null   object 
 13  Condition1    1460 non-null   object 
 14  Condition2    1460 non-null   object 
 15  BldgType      1460 non-null   object 
 16  HouseStyle    1460 non-null   object 
 17  OverallQual   1460 non-null   int64  
 18  OverallCond   1460 non-null 

In [8]:
#counting  nullvalues
print('df1 null values count')
print(f'LotFrontage : {df1.LotFrontage.isnull().sum()}')
print(f'Alley: {df1.Alley.isnull().sum()}')



df1 null values count
LotFrontage : 259
Alley: 1369


In [9]:
#showing the count of unique values of the Alley's variable
df1.Alley.value_counts()

Grvl    50
Pave    41
Name: Alley, dtype: int64

In our first dataframe it shows us 3 data types, (int64,object,float64), which tells us a bit about what we will be working with for our model. And specifically, there are two variables that stand out for the number of missing values they have, LotFrontage with 259 out of 1460 and Alley with 1369 out of 1460. The Alley variable tell us te type of the type of alley access, it can be  Gravel or  Paved, the problem is that it doesn't have enough information to be considered into our study, so it must be drop from the dataframe, this is the description of Alley's variable

Alley: Type of alley access to property

       Grvl	Gravel
       Pave	Paved
       NA 	No alley access
		

In [10]:
#Using the info function in our second dataframe
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   RoofStyle     1460 non-null   object 
 1   RoofMatl      1460 non-null   object 
 2   Exterior1st   1460 non-null   object 
 3   Exterior2nd   1460 non-null   object 
 4   MasVnrType    1452 non-null   object 
 5   MasVnrArea    1452 non-null   float64
 6   ExterQual     1460 non-null   object 
 7   ExterCond     1460 non-null   object 
 8   Foundation    1460 non-null   object 
 9   BsmtQual      1423 non-null   object 
 10  BsmtCond      1423 non-null   object 
 11  BsmtExposure  1422 non-null   object 
 12  BsmtFinType1  1423 non-null   object 
 13  BsmtFinSF1    1460 non-null   int64  
 14  BsmtFinType2  1422 non-null   object 
 15  BsmtFinSF2    1460 non-null   int64  
 16  BsmtUnfSF     1460 non-null   int64  
 17  TotalBsmtSF   1460 non-null   int64  
 18  Heating       1460 non-null 

In [11]:
print('df2 null values count')
print(f'MasVnrType : {df2.MasVnrType.isnull().sum()}')
print(f'MasVnrArea: {df2.MasVnrArea.isnull().sum()}')
print(f'BsmtQual: {df2.BsmtQual.isnull().sum()}')
print(f'BsmtCond: {df2.BsmtCond.isnull().sum()}')
print(f'BsmtExposure: {df2.BsmtExposure.isnull().sum()}')
print(f'BsmtFinType1: {df2.BsmtFinType1.isnull().sum()}')
print(f'BsmtFinType2: {df2.BsmtFinType2.isnull().sum()}')


df2 null values count
MasVnrType : 8
MasVnrArea: 8
BsmtQual: 37
BsmtCond: 37
BsmtExposure: 38
BsmtFinType1: 37
BsmtFinType2: 38



In the second dataframe there are several columns with missing values, the only difference with our previous dataframe is that in each column the amount of missing data is relatively small. The number of missing values ​​that are found is in a range of 8 and 38, so it would not make sense to remove any of these columns.

In [12]:
#Using the info function in our third dataframe
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 20 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Electrical    1459 non-null   object 
 1   1stFlrSF      1460 non-null   int64  
 2   2ndFlrSF      1460 non-null   int64  
 3   LowQualFinSF  1460 non-null   int64  
 4   GrLivArea     1460 non-null   int64  
 5   BsmtFullBath  1460 non-null   int64  
 6   BsmtHalfBath  1460 non-null   int64  
 7   FullBath      1460 non-null   int64  
 8   HalfBath      1460 non-null   int64  
 9   BedroomAbvGr  1460 non-null   int64  
 10  KitchenAbvGr  1460 non-null   int64  
 11  KitchenQual   1460 non-null   object 
 12  TotRmsAbvGrd  1460 non-null   int64  
 13  Functional    1460 non-null   object 
 14  Fireplaces    1460 non-null   int64  
 15  FireplaceQu   770 non-null    object 
 16  GarageType    1379 non-null   object 
 17  GarageYrBlt   1379 non-null   float64
 18  GarageFinish  1379 non-null 

In [13]:
print('df3 null values count')
print(f'Electrical: {df3.Electrical.isnull().sum()}')
print(f'FireplaceQu: {df3.FireplaceQu.isnull().sum()}')
print(f'GarageType: {df3.GarageType.isnull().sum()}')
print(f'GarageYrBlt: {df3.GarageYrBlt.isnull().sum()}')
print(f'GarageFinish: {df3.GarageFinish.isnull().sum()}')


df3 null values count
Electrical: 1
FireplaceQu: 690
GarageType: 81
GarageYrBlt: 81
GarageFinish: 81


In [14]:
#showing the count of unique values of the FireplaceQu's variable
df3.FireplaceQu.value_counts()

Gd    380
TA    313
Fa     33
Ex     24
Po     20
Name: FireplaceQu, dtype: int64


In the third dataframe created we also have variables with missing values, of which only one column is the one that stands out "FireplaceQu", this variable has a total of 690 missing values ​​out of 1460. FireplaceQu is the fireplace quality, the values that it has are the next ones:

       Ex	Excellent - Exceptional Masonry Fireplace,
       Gd	Good - Masonry Fireplace in main level,
       TA	Average - Prefabricated Fireplace in main living area or Masonry Fireplace in basement,
       Fa	Fair - Prefabricated Fireplace in basement,
       Po	Poor - Ben Franklin Stove,
       NA	No Fireplace.
       
There really is a significant amount of missing values, but there is the justification that there is no fireplace quality in the house, and in addition to this the amount of NA values ​​does not exceed even half of the total data, so the treatment that  will be  given to this variable and their missing values ​​will be to replace all NA values ​​with "NF"

		

In [15]:
#Using the info function in our fourth dataframe
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   GarageQual     1379 non-null   object
 1   GarageCond     1379 non-null   object
 2   PavedDrive     1460 non-null   object
 3   WoodDeckSF     1460 non-null   int64 
 4   OpenPorchSF    1460 non-null   int64 
 5   EnclosedPorch  1460 non-null   int64 
 6   3SsnPorch      1460 non-null   int64 
 7   ScreenPorch    1460 non-null   int64 
 8   PoolArea       1460 non-null   int64 
 9   PoolQC         7 non-null      object
 10  Fence          281 non-null    object
 11  MiscFeature    54 non-null     object
 12  MiscVal        1460 non-null   int64 
 13  MoSold         1460 non-null   int64 
 14  YrSold         1460 non-null   int64 
 15  SaleType       1460 non-null   object
 16  SaleCondition  1460 non-null   object
 17  SalePrice      1460 non-null   int64 
dtypes: int64(10), object(8)
memo

In [16]:
print('df4 null values count')
print(f'GarageQual: {df4.GarageQual.isnull().sum()}')
print(f'GarageCond: {df4.GarageCond.isnull().sum()}')
print(f'PoolQC: {df4.PoolQC.isnull().sum()}')
print(f'Fence: {df4.Fence.isnull().sum()}')
print(f'MiscFeature: {df4.MiscFeature.isnull().sum()}')

df4 null values count
GarageQual: 81
GarageCond: 81
PoolQC: 1453
Fence: 1179
MiscFeature: 1406


In [17]:
df4.PoolQC.value_counts()

Gd    3
Ex    2
Fa    2
Name: PoolQC, dtype: int64

In [18]:
df4.PoolArea.value_counts()

0      1453
480       1
512       1
519       1
555       1
576       1
648       1
738       1
Name: PoolArea, dtype: int64

In [19]:
df4.Fence.value_counts()

MnPrv    157
GdPrv     59
GdWo      54
MnWw      11
Name: Fence, dtype: int64

In [20]:
df4.MiscFeature.value_counts()

Shed    49
Othr     2
Gar2     2
TenC     1
Name: MiscFeature, dtype: int64

For the last Dataframe that was created, there are also missing values ​​where the most notable are in three variables ('PoolQC','Fence','MiscFeature'), each one has a significant number of missing values.The values ​​of each variable are shown as follows

PoolQC: Pool quality
		
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       NA	No Pool
		
Fence: Fence quality
		
       GdPrv	Good Privacy
       MnPrv	Minimum Privacy
       GdWo	Good Wood
       MnWw	Minimum Wood/Wire
       NA	No Fence
	
MiscFeature: Miscellaneous feature not covered in other categories
		
       Elev	Elevator
       Gar2	2nd Garage (if not described in garage section)
       Othr	Other
       Shed	Shed (over 100 SF)
       TenC	Tennis Court
       NA	None
       

When viewing and analyzing these variables, we only reach the conclusion of having to remove the 'PoolQC' and 'MiscFeature' columns from the dataframe, also because 'PoolArea' it is correlated to 'PoolQC' we would have to drop it to, since the amount of data they present is too low, in the case of the Fence variable, we prefer to modify the NA values ​​to NFe
       
       
       

 ### Drop unusable variables, treatment to missing values and creting a new database

In [21]:
# Drop variables
df_train.drop(['Id','Alley',"PoolQC","PoolArea","MiscFeature","GarageYrBlt"], axis =1 , inplace = True)

In [22]:
#Missing values treatment
df_train['LotFrontage'] = df_train['LotFrontage'].fillna('NLf')
df_train['MasVnrType'] = df_train['MasVnrType'].fillna('NMvt')
df_train['MasVnrArea'] = df_train['MasVnrArea'].fillna('0')
df_train['BsmtQual'] = df_train['BsmtQual'].fillna('NBsmt')
df_train['BsmtCond'] = df_train['BsmtCond'].fillna('NBsmtC')
df_train['BsmtExposure'] = df_train['BsmtExposure'].fillna('NBsmtE')
df_train['BsmtFinType1'] = df_train['BsmtFinType1'].fillna('NBsmt1')
df_train['BsmtFinType2'] = df_train['BsmtFinType2'].fillna('NBsmt2')
df_train['Electrical'] = df_train['Electrical'].fillna('')
df_train['FireplaceQu'] = df_train['FireplaceQu'].fillna('NFp')
df_train['GarageType'] = df_train['GarageType'].fillna('NG')
df_train['GarageFinish'] = df_train['GarageFinish'].fillna('NGf')
df_train['GarageQual'] = df_train['GarageQual'].fillna('NGq')
df_train['GarageCond'] = df_train['GarageCond'].fillna('NGc')
df_train['Fence'] = df_train['Fence'].fillna('NFe')

In [23]:
#Renaming columns
df_train.rename(columns={'1stFlrSF': 'FirstFlrSF', '2ndFlrSF': 'SecFlrSF','3SsnPorch': 'SsnPorch'},inplace=True)

1stFlrSF, 2ndFlrSF and 3SsnPorch columns are being replaced to avoid that column title starts with numeric digits.

In [24]:
#Saving the new database in a csv file

df_train.to_csv('df_ETL.csv', index = False)